File: frontend/src/lib/api/mealLogs.ts
import { z } from "zod";
import { API_URL } from "@/lib/config";
import { fetchWithAuth } from "@/lib/auth/client";

export const MEAL_SLOTS = ["breakfast", "morning_snack", "lunch", "evening_snack", "dinner"] as const;
export type MealSlot = (typeof MEAL_SLOTS)[number];

export const MEAL_SLOT_LABELS: Record<MealSlot, string> = {
  breakfast: "Breakfast",
  morning_snack: "Morning Snack",
  lunch: "Lunch",
  evening_snack: "Evening Snack",
  dinner: "Dinner",
};

export const MealLogOutSchema = z.object({
  id: z.string(),
  client_id: z.string(),
  hc_user_id: z.string(),
  meal_slot: z.enum(MEAL_SLOTS),
  description: z.string().nullable(),
  photo_original_filename: z.string(),
  photo_mime_type: z.string(),
  captured_at: z.string().nullable(),
  logged_at: z.string(),
  hc_reaction: z.enum(["happy", "neutral", "sad"]).nullable(),
  reacted_at: z.string().nullable(),
});
export type MealLogOut = z.infer<typeof MealLogOutSchema>;

const PaginatedMealLogsSchema = z.object({
  items: z.array(MealLogOutSchema),
  next_cursor: z.string().nullable(),
});

// ── HC-side ──────────────────────────────────────────────────────────────────

export async function listClientMealLogs(clientId: string): Promise<{ items: MealLogOut[]; next_cursor: string | null }> {
  const res = await fetchWithAuth(`${API_URL}/api/clients/${clientId}/meal-logs`);
  if (!res.ok) throw new Error(`List meal logs failed: ${res.status}`);
  return PaginatedMealLogsSchema.parse(await res.json());
}

export async function reactToMealLog(
  clientId: string, mealLogId: string, reaction: "happy" | "neutral" | "sad",
): Promise<MealLogOut> {
  const res = await fetchWithAuth(`${API_URL}/api/clients/${clientId}/meal-logs/${mealLogId}/react`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ reaction }),
  });
  if (!res.ok) throw new Error(`React to meal log failed: ${res.status}`);
  return MealLogOutSchema.parse(await res.json());
}

export function mealLogPhotoUrl(clientId: string, mealLogId: string): string {
  return `${API_URL}/api/clients/${clientId}/meal-logs/${mealLogId}/photo`;
}

// ── client-side ──────────────────────────────────────────────────────────────

export async function listMyMealLogs(): Promise<{ items: MealLogOut[]; next_cursor: string | null }> {
  const res = await fetchWithAuth(`${API_URL}/api/me/meal-logs`);
  if (!res.ok) throw new Error(`List my meal logs failed: ${res.status}`);
  return PaginatedMealLogsSchema.parse(await res.json());
}

export async function submitMyMealLog(input: { mealSlot: MealSlot; description?: string; photo: File }): Promise<MealLogOut> {
  const form = new FormData();
  form.append("meal_slot", input.mealSlot);
  if (input.description) form.append("description", input.description);
  form.append("photo", input.photo);

  const res = await fetchWithAuth(`${API_URL}/api/me/meal-logs`, { method: "POST", body: form });
  if (!res.ok) throw new Error(`Submit meal log failed: ${res.status}`);
  return MealLogOutSchema.parse(await res.json());
}

export function myMealLogPhotoUrl(mealLogId: string): string {
  return `${API_URL}/api/me/meal-logs/${mealLogId}/photo`;
}

File: backend/src/api/meal_logs.py
"""HC-side meal_logs list/react endpoints. Client-side submit lives in me.py."""
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from sqlalchemy import and_, or_, select

from src.api.deps import DbDep, HcClaimsDep, LimitDep, PaginatedList, TenantDep, decode_cursor, encode_cursor
from src.db.models import Client, MealLog

router = APIRouter(tags=["meal-logs"])

ALLOWED_MEAL_PHOTO_MIME_TYPES = {"image/jpeg", "image/png", "image/webp", "image/heic"}
MAX_MEAL_PHOTO_SIZE_BYTES = 10 * 1024 * 1024  # 10 MB — matches PHASE-02c's message-attachment cap
VALID_REACTIONS = {"happy", "neutral", "sad"}


# ── schemas ────────────────────────────────────────────────────────────────────


class MealLogOut(BaseModel):
    id: UUID
    client_id: UUID
    hc_user_id: UUID
    meal_slot: str
    description: str | None
    photo_original_filename: str
    photo_mime_type: str
    captured_at: datetime | None
    logged_at: datetime
    hc_reaction: str | None
    reacted_at: datetime | None

    model_config = {"from_attributes": True}


class MealLogReactIn(BaseModel):
    reaction: str


# ── shared helper (this module's own copy — see Global Constraints) ────────────


async def _get_owned_client(db: DbDep, client_id: UUID, hc_id: str) -> Client:
    client = (await db.execute(
        select(Client).where(Client.id == client_id, Client.hc_user_id == UUID(hc_id))
    )).scalar_one_or_none()
    if client is None:
        raise HTTPException(status_code=404, detail="Client not found")
    return client


# ── routes ─────────────────────────────────────────────────────────────────────


@router.get("/api/clients/{client_id}/meal-logs")
async def list_client_meal_logs(
    client_id: UUID,
    claims: HcClaimsDep,
    hc_id: TenantDep,
    db: DbDep,
    limit: LimitDep = 40,  # higher default than messages/check-ins — day-grouping wants more per page
    cursor: str | None = None,
) -> PaginatedList[MealLogOut]:
    await _get_owned_client(db, client_id, hc_id)

    q = select(MealLog).where(MealLog.client_id == client_id)
    if cursor:
        cur_ts, cur_id = decode_cursor(cursor)
        q = q.where(
            or_(
                MealLog.logged_at < cur_ts,
                and_(MealLog.logged_at == cur_ts, MealLog.id < cur_id),
            )
        )
    q = q.order_by(MealLog.logged_at.desc(), MealLog.id.desc()).limit(limit + 1)
    rows = (await db.execute(q)).scalars().all()

    next_cursor: str | None = None
    if len(rows) > limit:
        rows = rows[:limit]
        next_cursor = encode_cursor(rows[-1].logged_at, rows[-1].id)

    return PaginatedList(items=[MealLogOut.model_validate(r) for r in rows], next_cursor=next_cursor)


@router.post("/api/clients/{client_id}/meal-logs/{meal_log_id}/react")
async def react_to_meal_log(
    client_id: UUID,
    meal_log_id: UUID,
    body: MealLogReactIn,
    claims: HcClaimsDep,
    hc_id: TenantDep,
    db: DbDep,
) -> MealLogOut:
    await _get_owned_client(db, client_id, hc_id)

    if body.reaction not in VALID_REACTIONS:
        raise HTTPException(status_code=422, detail=f"reaction must be one of {sorted(VALID_REACTIONS)}")

    meal_log = (await db.execute(
        select(MealLog).where(MealLog.id == meal_log_id, MealLog.client_id == client_id)
    )).scalar_one_or_none()
    if meal_log is None:
        raise HTTPException(status_code=404, detail="Meal log not found")

    meal_log.hc_reaction = body.reaction
    meal_log.reacted_at = datetime.now(tz=meal_log.logged_at.tzinfo)

    await db.commit()
    await db.refresh(meal_log)
    return MealLogOut.model_validate(meal_log)
